File: LeetCode/[M] - 208. Implement Trie (Prefix Tree)/solution.go
package main

type Trie struct {
	isWord    bool
	childrens map[rune]*Trie
}

func Constructor() Trie {
	return Trie{childrens: make(map[rune]*Trie)}
}

func (t *Trie) Insert(word string) {
	for _, char := range word {
		if _, ok := t.childrens[char]; !ok {
			t.childrens[char] = &Trie{childrens: make(map[rune]*Trie)}
		}
		t = t.childrens[char]
	}
	t.isWord = true
}

func (t *Trie) Search(word string) bool {
	for _, char := range word {
		if _, ok := t.childrens[char]; !ok {
			return false
		}
		t = t.childrens[char]
	}
	return t.isWord
}

func (t *Trie) StartsWith(prefix string) bool {
	for _, char := range prefix {
		if _, ok := t.childrens[char]; !ok {
			return false
		}
		t = t.childrens[char]
	}
	return true
}

/**
 * Your Trie object will be instantiated and called as such:
 * obj := Constructor();
 * obj.Insert(word);
 * param_2 := obj.Search(word);
 * param_3 := obj.StartsWith(prefix);
 */

File: LeetCode/[M] - 208. Implement Trie (Prefix Tree)/solution.py
class Trie:
    def __init__(self):
        self.trie = {}

    def insert(self, word: str) -> None:
        node = self.trie
        for ch in word:
            if ch not in node:
                node[ch] = {"is_word": False}
            node = node[ch]
        node["is_word"] = True

    def search(self, word: str) -> bool:
        node = self.trie
        for ch in word:
            if ch not in node:
                return False
            node = node[ch]
        return node["is_word"]

    def startsWith(self, prefix: str) -> bool:
        node = self.trie
        for ch in prefix:
            if ch not in node:
                return False
            node = node[ch]
        return True


# Your Trie object will be instantiated and called as such:
# obj = Trie()
# obj.insert(word)
# param_2 = obj.search(word)
# param_3 = obj.startsWith(prefix)
